### PR TITLE
fix(manager/gitlabci): strip Gitlab Dependecy Proxy variable from depName

### LINKED ISSUE #9958: Support Gitlab Dependency Proxy Images
**What would you like Renovate to be able to do?**

Gitlab recently implemented a dependency proxy to work around the dockerhub rate limitation.
The dependency proxy basically acts as a cache that only contacts dockerhub if it does not have a local image or the image is stale. (See also the documentation here https://docs.gitlab.com/ee/user/packages/dependency_proxy/#authenticate-within-cicd)

To use this dependency proxy within the `.gitlab-ci.yml` you specify something like

```
build:
  image: ${CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX}/docker:20.10.5
  stage: build
  script:
    - echo "do something great"
```

notice the `${CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX}` in front of the image name. 
This variable causes renovate to skip this image entry.

```
{
 "skipReason": "contains-variable",
 "replaceString": "${CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX}/docker:20.10.5",
 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
 "datasource": "docker",
 "depType": "image-name",
 "depIndex": 0,
 "updates": []
},
```

It would be great if renovate would ignore this specific environment variable during the new version check.
However it would be important that it does not remove it in the suggested MergeRequest.

**Did you already have any implementation ideas?**

I would assume that the implementation would happen in https://github.com/renovatebot/renovate/blob/main/lib/manager/gitlabci/extract.ts by wrapping `getDep` and stripping the variable before the call. I do however not know how to ensure the variable is put back on after a change.

If you would need someone to test an implementation for this feature I would be glad to help.

Thank you

### PR DESCRIPTION
**Pull request by dsch**
## Changes

Remove image prefix variable from images pulled through the Dependency Proxy.

## Context

The predefined variables `CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX` and `CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX` can be used to pull images from Docker Hub through Gitlab Dependency Proxy. Since the part after the prefix is a resolvable image name, only this part use used as `depName`. Also the `replaceString` does not contain the variable, since it's not relevant for replacement.

Closes #9958

_Note:_ This change does not solve the two additionally mentioned use-cases in #9958: [CI_REGISTRY](https://github.com/renovatebot/renovate/issues/9958#issuecomment-919866846) and [Dependency Proxy in Dockerfiles](https://github.com/renovatebot/renovate/issues/9958#issuecomment-954967674).

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by JamieMagee** [comment]
@dsch Overall your change looks good. I had a go at resolving the conflicts with `main`, but it wasn't straightforward 😅

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs deconflicting

**Comment by dsch** [comment]
@rarkins thank you for marking the PR as draft. 
I'm kind of blocked on this: The conflict was introduced in 6558d051f4d27014d42d57f1b89d91ab0667067b which causes issue #14765 and makes the newly created test cases fail.
I'm considering two option: 
- Weakening my testcase so they don't fail
- rewriting the gitlab-manager by using `js-yaml`

File: tsconfig.strict.json [hunk 1]
```diff
@@ -156,6 +156,7 @@
     "lib/modules/manager/gitlabci-include/index.ts",
     "lib/modules/manager/gitlabci/extract.ts",
     "lib/modules/manager/gitlabci/index.ts",
+    "lib/modules/manager/gitlabci/utils.ts",
```
Comment (JamieMagee) [inline tsconfig.strict.json:159]: Required because of the new imports in `utils.ts` which are not strictly checked (yet!)

**Review by JamieMagee — DISMISSED** [review]

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/gitlabci/utils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import { getGitlabDep } from './utils';
+
+describe('modules/manager/gitlabci/utils', () => {
+  describe('getGitlabDep', () => {
+    it.each([
```
Comment (viceice) [inline lib/modules/manager/gitlabci/utils.spec.ts:5]: can we use the string template table syntax instead, it's much more readable. checkout versioning tests for samples.

Comment (dsch) [reply lib/modules/manager/gitlabci/utils.spec.ts]: @viceice could please have another look at this PR. Tests use now string template table syntax.

Comment (viceice) [reply lib/modules/manager/gitlabci/utils.spec.ts]: Please 👉 https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

**Comment by viceice** [comment]
Your latest commit is unverified 😕 

https://github.com/renovatebot/renovate/pull/14734/commits/0c4ac204dd95d80e2428caf459ecd12f1d99b6ca

File: lib/modules/manager/gitlabci/extract.spec.ts [hunk 3]
```diff
@@ -160,6 +160,15 @@ describe('modules/manager/gitlabci/extract', () => {
         },
       ]);
     });
+
+    it('extract images from dependency proxy', async () => {
+      const res = await extractAllPackageFiles(config, [
+        'lib/modules/manager/gitlabci/__fixtures__/gitlab-ci.8.yaml',
+      ]);
+      expect(res).toMatchSnapshot();
+      expect(res).toHaveLength(1);
```
[the remark this answers is not in the document]
Comment (dsch) [reply lib/modules/manager/gitlabci/extract.spec.ts]: I used `expect.toMatchInlineSnapshot()` in 6aecd07b0fc432bd44bbf2b9ed6f57130fb70cfa, but I'm not sure if this makes it really more readable.

**Review by viceice — COMMENTED** [review]
needs deconflicting

File: lib/modules/manager/gitlabci/extract.spec.ts [hunk 4]
```diff
@@ -167,6 +167,43 @@ describe('modules/manager/gitlabci/extract', () => {
...
+            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
+            "currentDigest": undefined,
+            "currentValue": "1.0.0",
+            "datasource": "docker",
+            "depName": "other/image1",
+            "depType": "service-image",
+            "replaceString": "other/image1:1.0.0",
+          },
+        ]
+      `);
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.spec.ts:204]: Please convert to `toEqual`
```suggestion
      expect(res[0].deps).toEqual(
        [
          {
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "currentDigest": undefined,
            "currentValue": "31.65.1-slim",
            "datasource": "docker",
            "depName": "renovate/renovate",
            "depType": "image-name",
            "replaceString": "renovate/renovate:31.65.1-slim",
          },
          {
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "currentDigest": undefined,
            "currentValue": "10.4.11",
            "datasource": "docker",
            "depName": "mariadb",
            "depType": "service-image",
            "replaceString": "mariadb:10.4.11",
          },
          {
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "currentDigest": undefined,
            "currentValue": "1.0.0",
            "datasource": "docker",
            "depName": "other/image1",
            "depType": "service-image",
            "replaceString": "other/image1:1.0.0",
          },
        ]
      );
```

File: lib/modules/manager/gitlabci/utils.ts [hunk 5]
```diff
@@ -10,3 +14,21 @@ export function replaceReferenceTags(content: string): string {
...
+
+/**
+ * Get image dependencies respecting Gitlab Dependency Proxy
+ * @param imageName as used in .gitlab-ci.yml file
+ * @return package dependency for the image
+ */
+export function getGitlabDep(imageName: string): PackageDependency {
+  const match = depProxyRe.exec(imageName);
+  if (match?.groups) {
+    return getDep(match.groups.depName);
```
Comment (viceice) [inline lib/modules/manager/gitlabci/utils.ts:30]:
```suggestion
    return { ...getDep(match.groups.depName), replaceString: imageName ];
```

I would prefer to have the full name for replacement, so it's more unique for auto-replace.

@rarkins WDYT?

File: lib/modules/manager/gitlabci/__fixtures__/gitlab-ci.8.yaml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+image:
+  # comment
+  name: ${CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX}/renovate/renovate:31.65.1-slim
+
+services:
+  # comment
+  - $CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX/mariadb:10.4.11
+  # another comment
+  - name: $CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX/other/image1:1.0.0
+    alias: imagealias1
```
Comment (viceice) [inline lib/modules/manager/gitlabci/__fixtures__/gitlab-ci.8.yaml:10]: Please inline this small snipet

**Comment by viceice** [comment]
https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

File: lib/modules/manager/gitlabci/utils.ts (hunk 5 above)
Comment (viceice) [reply lib/modules/manager/gitlabci/utils.ts:30]: @rarkins ☝️

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/manager/gitlabci/utils.ts (hunk 5 above)
Comment (rarkins) [reply lib/modules/manager/gitlabci/utils.ts:30]: Yes, that would be better

**Comment by dsch** [comment]
By using the full name for replacement, the prefix gets dropped in the PR.
I marked the PR as draft until this is resolved.

image

**Comment by rarkins** [comment]
@viceice This may be hard to fix due to the way autoReplace works, unless we manually specify the replace template during extract. Shall we revert to how it was?

**Comment by viceice** [comment]
yes, let's revert. sorry

**Comment by dsch** [comment]
@viceice I implemented a prefixed version of autoReplaceTemplateString. I'm not sure if this goes in the right direction, because it still does not work as expected. 
Any hint where I can learn about autoReplace?

**Comment by dsch** [comment]
After deleting the branch from my prevision test it work.
https://gitlab.com/dsch1/renovate-gitlab-dependency-proxy/-/merge_requests/3

File: lib/modules/manager/gitlabci/utils.ts [hunk 7]
```diff
@@ -10,3 +14,24 @@ export function replaceReferenceTags(content: string): string {
   const res = content.replace(re, '');
   return res;
 }
+
+const depProxyRe = regEx(
+  `(?<prefix>\\$\\{?CI_DEPENDENCY_PROXY_(?:DIRECT_)?GROUP_IMAGE_PREFIX\\}?\\/)?(?<depName>.+)`
```
Comment (viceice) [inline lib/modules/manager/gitlabci/utils.ts:19]:
```suggestion
  `(?<prefix>\\$\\{?CI_DEPENDENCY_PROXY_(?:DIRECT_)?GROUP_IMAGE_PREFIX\\}?\\/)(?<depName>.+)`
```
shouldn't optional, as that will catched by default.

File: lib/modules/manager/gitlabci/utils.ts [hunk 8]
```diff
@@ -10,3 +14,24 @@ export function replaceReferenceTags(content: string): string {
...
+ * Get image dependencies respecting Gitlab Dependency Proxy
+ * @param imageName as used in .gitlab-ci.yml file
+ * @return package dependency for the image
+ */
+export function getGitlabDep(imageName: string): PackageDependency {
+  const match = depProxyRe.exec(imageName);
+  if (match?.groups) {
+    const dep = { ...getDep(match.groups.depName), replaceString: imageName };
+    dep.autoReplaceStringTemplate =
+      (match.groups.prefix ?? '') + dep.autoReplaceStringTemplate;
```
Comment (viceice) [inline lib/modules/manager/gitlabci/utils.ts:32]:
```suggestion
      (match.groups.prefix) + dep.autoReplaceStringTemplate;
```
shouldn't optional.

File: lib/modules/manager/gitlabci/utils.spec.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      ).toMatchObject({
+        autoReplaceStringTemplate: defaultAutoReplaceStringTemplate,
+        replaceString: 'quay.io/prometheus/node-exporter:v1.3.1',
+        depName: 'quay.io/prometheus/node-exporter',
+        currentValue: 'v1.3.1',
+      });
+    });
+
+    it('empty', () => {
+      expect(getGitlabDep('')).toMatchObject({
```
Comment (viceice) [inline lib/modules/manager/gitlabci/utils.spec.ts:57]: i think this should have a skip reason, as it's invalid

File: lib/modules/manager/gitlabci/utils.spec.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    it.each`
+      name                           | imagePrefix
+      ${'no variable'}               | ${''}
+      ${'group proxy'}               | ${'$CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX/'}
+      ${'group proxy with brackets'} | ${'${CI_DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX}/'}
+      ${'direct group proxy'}        | ${'$CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX/'}
+    `(
+      'image with organization - $name',
+      ({ imagePrefix }: { imagePrefix: string }) => {
+        const imageName = imagePrefix + 'renovate/renovate:19.70.8-slim';
```
Comment (viceice) [inline lib/modules/manager/gitlabci/utils.spec.ts:34]: please use string templates

**Review by viceice — COMMENTED** [review]
i don't see failing tests yet 🙃

**Review by viceice — APPROVED** [review]

**Comment by dsch** [comment]
@viceice thank you for supporting me on my endeavor to create my first TypeScript PR.

**Comment by rarkins** [comment]
@dsch most of all, thanks to you!

**Pull request merged**